### PR TITLE
Off file mesh component

### PR DESCRIPTION
**Pull request by Lomholy**
### Free-form text area
_Please describe what your PR is adding in terms of features or bugfixes:_
This pull requests changes the Union_mesh.comp to include .off files, and both the filetypes that .stl files come in (binary and ascii).

It does this through changes in the component itself, but also through changes in union-lib.c and the instrument test file Geometry_test.instr.

--------------
### Development OS / boundary conditions
_Please describe what OS you developed and tested your additions on, and if any special dependencies are required:_
I developed on macos 26.1 Tahoe, and no of no special dependencies. Has also been tested on an Ubuntu 24.04.3 ARM.

--------------
# PR Checklist for contributing to McStas/McXtrace
## For a coherent and useful contribution to McStas/McXtrace, please fill in _relevant parts_ of the checklist:
* ### My contribution includes patches to an **existing component** file
  * [X] I have used the `mcdoc` utility and **rendered** a reasonable documentation page for the component (please attach as screenshot in comments!)
  * [X] I have ensured that basic use of the component is OK (e.g. an instrument using it compiles?)
  * [X] I have used the `mctest` utility to **test** one or more instruments making use of the component (please attach `mcviewtest` report as screenshot in comments)
  * [X] I have used the `mcrun --c-lint` "linter" and followed advice to remove most / all warnings that are raised
* ### My contribution includes patches to an **existing** instrument file
  * [X] I have used the `mcdoc` utility and **rendered** a reasonable documentation page for the instrument (please attach as screenshot in comments!)
  * [X] I have used the `mctest` utility to **test** the instrument (please attach `mcviewtest` report as screenshot in comments)
  * [X] I have used the `mcrun --c-lint` "linter" and followed advice to remove most / all warnings that are raised
* ### My contribution includes a **new component** file
  * [ ] I have ensured that naming of parameters are in the style of existing components. (Please check the [McStas](https://github.com/mccode-dev/McCode/blob/main/mcstas-comps/NOMENCLATURE.md) or [McXtrace](https://github.com/mccode-dev/McCode/blob/main/mcxtrace-comps/NOMENCLATURE) NOMENCLATURE docs.)
  * [ ] I have ensured that component parameters are in the usually units of McStas or McXtrace (SI + neutron/x-ray 'usual' units)
  * [ ] I have used the `mcdoc` utility and **rendered** a reasonable documentation page for the component (please attach as screenshot in comments!)
  * [ ] I have ensured that basic use of the component is OK (e.g. an instrument using it compiles?)
  * [ ] I have included a corresponding **example** instrument and will fill in the **new instrument** section below
  * [ ] My new component is added within the `contrib` component category
* ### My contribution includes a **new instrument** file
  * [ ] I have used the `mcdoc` utility and **rendered** a reasonable documentation page for the instrument (please attach as screenshot in comments!)
  * [ ] I have ensured that basic use of the instrument is OK (e.g. it compiles?)
  * [ ] ... and provided reasonable default parameters in that instrument that produce reasonable output
  * [ ] ... and maybe even added a `%Example:` line to describe expected behaviour
  * [ ] I have used the `mcrun --c-lint` "linter" and followed advice to remove most / all warnings that are raised
  * [ ] My new instrument is added within the `examples` hierarchy in a folder in the style of `examples/ESS/New_stuff/New_stuff.instr`
  * [ ] My new instrument has a new, unique filename, not clashing with existing example instruments
  * [ ] My new instrument requires a data/input file. If the datafile is _specific_ for my instrument I have left it in the same `example` folder, but if general use I have placed it in the global `data` folder.
* ### My work touches the code-generator in mccode/src
  * [ ] I have added reasoning and documentation for the change through an ADR record in our [GRAMMAR](https://github.com/mccode-dev/McCode/tree/main/docs/GRAMMAR) section
  * [ ] I am attaching test output in the comments
* ### My work touches / adds to the runtime lib code (.c,.h etc in multiple locations
  * [ ] I am have added reasoning and documentation for the change below
  * [ ] I am attaching test output in the comments
* ### My PR is meant to fix a specific, existing issue
  * [ ] I have indicated the issue number here:
  * [ ] I have added documentation for the fix and possible side effects
* ### My contribution contains something else
  * [ ] Explanation is added in free form text above or below the checklist

--------------

### CODE REVIEW

**Comment by Lomholy** [comment]
--c-lint did not pass well, although this seemed to be due to other code sites in the Union framework.

Here is the mcdoc for Geometry_test.instr:
Screenshot 2025-11-27 at 16 38 19

Here is the mcdoc for the Union_mesh.comp:

Screenshot 2025-11-27 at 16 32 19

Here is the mcviewtest for the test instrument:

Screenshot 2025-11-27 at 16 30 22

**Comment by willend** [comment]
@Lomholy the fix for Windows/MSVC was easy, one needs to use malloc / calloc in place of [ ] style allocation with MSVC. [ ] can more or less only be used with literal, constant sizes.

**Comment by Lomholy** [comment]
@willend @mads-bertelsen  I have just implemented Mads' suggested change to the best of my abilities in my local work directory, and it seems like this provoked the issue that we are seeing in windows is now occuring. Will work on and see if I can find the issue.

File: mcstas-comps/union/Union_master.comp [hunk 1]
```diff
@@ -1265,7 +1265,7 @@ TRACE
               printf("Checking if child %d of current_volume = %d have intersections. \n",*check,current_volume);
             #endif
             // Only check the first of the two results in the intersection table, as they are ordered, and the second is of no interest
-            if (intersection_time_table.calculated[*check] == 1 && intersection_time_table.intersection_times[*check][0] > time_propagated_without_scattering) {
+            if (intersection_time_table.calculated[*check] == 1 && intersection_time_table.intersection_times[*check][0] > time_propagated_without_scattering && !Volumes[current_volume]->skip_hierarchical_optimization) {
```
Comment (mads-bertelsen) [inline mcstas-comps/union/Union_master.comp]: Here the information skip_hierachical_optimization is placed on the volume structure, since it's related to the geometry, I think it's nicer to place it on the geometry structure. This is also easier to access in the geometry components.

The check here in master can be moved outside of the for loop as it doesn't relate to the children's of the current volume, but just the volume itself, so it can just be an if from line 1262 to 1304, which would prevent intersection_with_children being set to 1 and thus the intersection call will be made.

File: mcstas-comps/union/Union_mesh.comp [hunk 2]
```diff
@@ -65,235 +66,516 @@
...
+      (*faces)[i][2] = j+2;
+   }
+   free(triangles);
+   // Loop over vertices, and make sure we only use unique vertices
+   Coords* unique_vertices = (Coords *)malloc(
+                  sizeof(Coords)* *n_verts) ;
+ 
+   int vertex_is_unique;
+   int x_are_equal, y_are_equal, z_are_equal;
+   int* map_old_to_unique = malloc(sizeof(int)* *n_verts);
```
Comment (mads-bertelsen) [inline mcstas-comps/union/Union_mesh.comp]: I don't believe this list is free'ed after use.

File: mcstas-comps/union/Union_mesh.comp [hunk 3]
```diff
@@ -65,235 +66,516 @@
...
+      (*verts)[j+2].y = (double)triangles[i].vertex3[1];
+      (*verts)[j+2].z = (double)triangles[i].vertex3[2];
+      (*faces)[i][0] = j;
+      (*faces)[i][1] = j+1;
+      (*faces)[i][2] = j+2;
+   }
+   free(triangles);
+   // Loop over vertices, and make sure we only use unique vertices
+   Coords* unique_vertices = (Coords *)malloc(
+                  sizeof(Coords)* *n_verts) ;
```
[the remark this answers is not in the document]
Comment (Lomholy) [reply mcstas-comps/union/Union_mesh.comp]: Freed this one, and the one above now

File: mcstas-comps/share/union-lib.c [hunk 4]
```diff
@@ -3878,25 +3879,45 @@ int sample_mesh_intersect(double *t,int *num_solutions,double *r,double *v,struc
...
             t[counter2] = t_intersect[iter];
-            counter2++;
+            x_intersect = t[counter2]*v[0] + x_new;
+		      y_intersect = t[counter2]*v[1] + y_new;
+		      z_intersect = t[counter2]*v[2] + z_new;
+            coordinates = coords_set(x_intersect,y_intersect,z_intersect);
+		      NORM(coordinates.x, coordinates.y, coordinates.z);
+		      nx[counter2] = coordinates.x;
+		      ny[counter2] = coordinates.y;
+		      nz[counter2] = coordinates.z;
```
Comment (mads-bertelsen) [inline mcstas-comps/share/union-lib.c]: nx ny and nz are the components of the normal vector for the intersected surface, but they are just provided the coordinates for the intersection instead. This might be inspired by the sphere case, as that happens to work there. In this case we instead need to grab the normal data in the mesh_storage with the appropriate index.

**Comment by Lomholy** [comment]
@willend Sounds like a good plan. I will gladly help starting to fix some of the linter warnings. Should I start by cherry picking over the commits? Or will you do this?

**Comment by willend** [comment]
@Lomholy doing **deep** debugging with @mads-bertelsen we found that for some reason, only ~1/8th of the torus is read on Windows. The issue seems to relate to the binary STL format only.

I've successfully run with an off-based geometry instead, including that model and a small modification to the instr in this regard. Chances are that tests run through now...

(Here is a screenshot from my windows machine - someone has cut the donut!)
Screenshot 2025-12-04 at 16 21 47

**Comment by willend** [comment]
It might be that we also need the patch that just arrived on https://github.com/mccode-dev/McCode/pull/2227, at least a local Windows install seems to indicate that. Will cherry-pick later.

**Comment by willend** [comment]
@Lomholy I've merged in my linter-work from the earlier-mentioned [neighbouring branch](https://github.com/mccode-dev/McCode/tree/off_file_mesh_MSVC_linter_patches) 

As far as I can see, the remaining linter-oriented warnings/errors relate only/mostly to the mesh-code from `union-lib`:
[linter.txt](https://github.com/user-attachments/files/23959246/linter.txt)

**Comment by willend** [comment]
No rush @Lomholy - enjoy the new offices :-)
